File: C++/NPC AI work/Finite State Machine/EnemyAI/StateMachine.cpp

#include "EnemyAI/StateMachine.h"


// Sets default values for this component's properties
UStateMachine::UStateMachine()
{
}

/// <summary>
/// Functions as the tick/update method for the various states
/// </summary>
void UStateMachine::SMTick() {
	Transition* transition = GetTransition();

	if (transition != nullptr) {
		SetState(transition->To);
	}

	if (currentState != nullptr) {
		currentState->GetInterface()->SMTick();
	}
}

/// <summary>
/// Sets a new state and does so in the following order:
/// 1. Runs the exit code for the current state
/// 2. Sets the new state
/// 3. Adds the TArray of transitions that apply to the new state to the currentState variable
/// 4. Runs the enter code for the new state
/// </summary>
/// <param name="state">The state that will be transitioned to. Takes a pointer to A UObject that inherits from the IState interface. </param>
void UStateMachine::SetState(TScriptInterface<IState>* state)
{
	if (state == currentState) {
		return;
	}

	if (currentState != nullptr) {
		currentState->GetInterface()->OnExit();
	}

	currentState = state;

	//Get the list of transitions for the current state and add it to the current transition
	currentTransitions = transitions.Find(currentState->GetObject()->GetClass());

	currentState->GetInterface()->OnEnter();

}

/// <summary>
/// Adds a transition clause between two defined states.
/// 
/// The states take pointers to A UObject that inherits from the IState interface.
/// </summary>
/// <param name="from">The state that will be transitioned out of.</param>
/// <param name="to">The state that will be transitioned to.</param>
/// <param name="predicate">A Tfunction that returns a boolean, this will be what is checked to decide if the transition can occur.</param>
void UStateMachine::AddTransition(TScriptInterface<IState>* from, TScriptInterface<IState>* to, TFunction<bool()> predicate)
{
	UClass* fromClass = from->GetObject()->GetClass();

	if (!transitions.Contains(fromClass)) {
		transitions.Emplace(fromClass);
	}

	TArray<UStateMachine::Transition>& currList = transitions[fromClass];


	// Create a transition and add it to the array

	currList.Add(UStateMachine::Transition(to, predicate));

}

/// <summary>
/// Adds a transition that can be done from any state.
/// </summary>
/// <param name="state">The state that will be transitioned to. Takes a pointer to A UObject that inherits from the IState interface.</param>
/// <param name="predicate">A Tfunction that returns a boolean, this will be what is checked to decide if the transition can occur.</param>
void UStateMachine::AddAnyTransition(TScriptInterface<IState>* state, TFunction<bool()> predicate)
{
	anyTransitions.Emplace(UStateMachine::Transition(state, predicate));
}

/// <summary>
/// Checks if any valid transitions can occur and returns the first valid one found.
/// 
/// Note: an AnyTransition will always take precedence over a transition between two specified states
/// </summary>
/// <returns></returns>
UStateMachine::Transition* UStateMachine::GetTransition()
{
	for (Transition& transition : anyTransitions) {
		if (transition.Condition()) {
			return &transition;
		}
	}

	//if there are no valid anyTransitions then we will get here

	for (Transition& transition : *currentTransitions) {
		if (transition.Condition()) {
			return &transition;
		}
	}

	return nullptr;
}

File: Infiltrator/Enemy State Machine/StateMachine.cs
using System;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class StateMachine
{
    // The state actively being run
    private IState currentState;
    //Dictionaries are basically maps
    // a map of lists containing different transitions based on the type they belong to 
    private Dictionary<Type, List<Transition>> transitions = new Dictionary<Type, List<Transition>>();
    private List<Transition> currentTransitions = new List<Transition>();
    // Transitions coming from any state, this is never switched out
    private List<Transition> anyTransitions = new List<Transition>();
    // An empty list for quickly creating a new transition list
    private static List<Transition> emptyTransitions = new List<Transition>(capacity:0);


    public void tick() {
        Transition transition = GetTransition();

        if(transition != null) {
            SetState(transition.To);
        }

        currentState?.Tick();
    }

    public void SetState(IState state) {
        if (state == currentState) {
            return;
        }
        
        currentState?.OnExit();
        currentState = state;

        // Get the list of transitions for the current state and add it to the current transition
        transitions.TryGetValue(currentState.GetType(), out currentTransitions);
        if(currentTransitions == null) {
            // If there was nothing make a new list
            currentTransitions = emptyTransitions;
        }
        currentState.OnEnter();
        
    }

    public void AddTransition(IState from, IState to, Func<bool> predicate) {

        if(transitions.TryGetValue(from.GetType(), out var currList) == false) {
            // If there is no current list for the from transition then make a new one
            // then add it to the dictionary
            currList = new List<Transition>();
            transitions.Add(from.GetType(), currList);
        }

        // Create a transition and add it to the list
        currList.Add(new Transition(to, predicate));
    }

    public void AddAnyTransition(IState state, Func<bool> predicate) {
        anyTransitions.Add(new Transition(state, predicate));
    }

    private class Transition {
        public Func<bool> Condition;
        public IState To;

        public Transition(IState to, Func<bool> condition) {
            To = to;
            Condition = condition;
        }
    }

    private Transition GetTransition() {

        foreach(Transition transition in anyTransitions) {
            if(transition.Condition()) {
                return transition;
            }
        }

        // If there are no valid anyTransitions then you will get here
        foreach(Transition transition in currentTransitions) {
            if(transition.Condition()) {
                return transition;
            }
        }

        return null;
    }
}
